cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -259,25 +289,93 @@ public void visitEnd() {
+                                        entityPartType.getClassName(), enforceOn.getClassName(), methodName, className);
+            Preconditions.checkArgument(getRequiredSize(enforceOn) != 0, "Invalid enforceOn %s provided.",
+                                        enforceOn.getClassName());
+          } else {
+            // Since the entity part/parts provided is of String type so validate that we have sufficient part for
+            // EntityId creation
+            // verify that all parts are of type string
+            for (EntityPartDetail entityPartDetail : entityPartDetails) {
+              entityPartType = getEntityPartType(parameterAnnotationNode, paramAnnotation, entityPartDetail);
+              Preconditions.checkArgument(entityPartType.equals(Type.getType(String.class)),
+                                          "Found part %s of type %s in a multiple part entity specification of " +
+                                            "AuthEnforce. Only String is supported in multiple parts.",
+                                          entityPartDetail.getEntityName(), entityPartType);
+            }
+            int requiredSize = getRequiredSize(enforceOn);
+            Preconditions.checkArgument(requiredSize != 0, "Failed to determine required number of entity parts " +
+              "needed for %s. Please make sure its a valid %s class for authorization enforcement",
+                                        enforceOn.getClassName(), EntityId.class.getSimpleName());
+            Preconditions.checkArgument(requiredSize == entityPartDetails.size(), "Found %s entity parts in " +
+                                          "AuthEnforce annotation on method %s in class %s to do enforcement on %s " +
+                                          "which requires %s entity parts or an %s", entityPartDetails.size(),
+                                        methodName, className, enforceOn, requiredSize, enforceOn.getClassName());
+
+          }
+        }
+
+        private Type getEntityPartType(Map<Integer, ParameterDetail> parameterAnnotationNode,
+                                       Map<String, Integer> paramAnnotation, EntityPartDetail entityPartDetail) {
+          return entityPartDetail.isField() ? fieldDetails.get(entityPartDetail.getEntityName()) :
+            parameterAnnotationNode.get(paramAnnotation.get(entityPartDetail.getEntityName())).getParameterType();
+        }
+
+        /**
+         * Return the required size of entity parts to create the {@link EntityId} on which authorization enforcement
+         * needs to be done as specified in {@link AuthEnforce#enforceOn()}
+         *
+         * @param enforceOn the {@link Type} of {@link EntityId} on which enforcement needs to be done
+         * @return the size of entity parts needed to create the above {@link EntityId} or 0 if found an invalid
+         * {@link EntityId} type.
+         */
+        private int getRequiredSize(Type enforceOn) {
```

Comment:
Should move this into the `AuthEnforceUtil` and leverage the `CONS_CACHE` (see below) to do this check.